File: squire-backend/api/routes/analytics.js
// Analytics and Insights Routes
const express = require('express');
const { createClient } = require('@supabase/supabase-js');
const router = express.Router();

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_ANON_KEY
);

// GET /api/analytics/dashboard/:userId - Get comprehensive dashboard data
router.get('/dashboard/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const { days = 30 } = req.query;

    const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();

    // Get all data in parallel
    const [
      { data: sessions },
      { data: suggestions },
      { data: events },
      { data: nodes },
      { data: relationships }
    ] = await Promise.all([
      supabase.from('user_sessions').select('*').eq('user_id', userId).gte('created_at', startDate),
      supabase.from('ai_suggestions').select('*').eq('user_id', userId).gte('created_at', startDate),
      supabase.from('user_events').select('*').eq('user_id', userId).gte('created_at', startDate),
      supabase.from('knowledge_nodes').select('*').eq('user_id', userId),
      supabase.from('knowledge_relationships').select('*').eq('user_id', userId)
    ]);

    const dashboard = {
      overview: {
        total_sessions: sessions?.length || 0,
        active_sessions: sessions?.filter(s => !s.session_end).length || 0,
        total_suggestions: suggestions?.length || 0,
        pending_suggestions: suggestions?.filter(s => s.status === 'pending').length || 0,
        total_events: events?.length || 0,
        knowledge_nodes: nodes?.length || 0,
        knowledge_relationships: relationships?.length || 0
      },
      trends: {
        daily_sessions: this.getDailyCounts(sessions || [], days),
        daily_events: this.getDailyCounts(events || [], days),
        suggestion_response_rate: suggestions?.length ?
          suggestions.filter(s => s.status !== 'pending').length / suggestions.length : 0
      },
      productivity: {
        avg_session_duration: this.calculateAvgSessionDuration(sessions || []),
        most_active_hours: this.getMostActiveHours(sessions || []),
        top_event_types: this.getTopEventTypes(events || []),
        knowledge_growth: this.calculateKnowledgeGrowth(nodes || [])
      },
      ai_insights: {
        suggestion_accuracy: this.calculateSuggestionAccuracy(suggestions || []),
        top_suggestion_types: this.getTopSuggestionTypes(suggestions || []),
        confidence_distribution: this.getConfidenceDistribution(suggestions || [])
      }
    };

    res.json({ dashboard });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/analytics/session-insights/:sessionId - Get detailed session insights
router.get('/session-insights/:sessionId', async (req, res) => {
  try {
    const { sessionId } = req.params;

    // Get session and user info
    const { data: session, error: sessionError } = await supabase
      .from('user_sessions')
      .select('*')
      .eq('id', sessionId)
      .single();

    if (sessionError) throw sessionError;
    if (!session) return res.status(404).json({ error: 'Session not found' });

    const { data, error } = await supabase
      .rpc('generate_session_insights', {
        p_user_id: session.user_id,
        p_session_id: sessionId
      });

    if (error) throw error;

    // Add additional analytics
    const insights = {
      ...data,
      ocr_analysis: this.analyzeOCRData(session.ocr_logs || []),
      click_patterns: this.analyzeClickPatterns(session.clicks || []),
      mouse_patterns: this.analyzeMousePatterns(session.mouse_movements || []),
      productivity_score: this.calculateProductivityScore(session)
    };

    res.json({ insights });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/analytics/behavior-patterns/:userId - Get behavioral patterns
router.get('/behavior-patterns/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const { days = 30, pattern_type } = req.query;

    const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();

    // Get events for pattern analysis
    const { data: events, error } = await supabase
      .from('user_events')
      .select('*')
      .eq('user_id', userId)
      .gte('created_at', startDate)
      .order('created_at');

    if (error) throw error;

    const patterns = {
      temporal: this.analyzeTemporalPatterns(events || []),
      frequency: this.analyzeFrequencyPatterns(events || []),
      sequences: this.analyzeSequencePatterns(events || []),
      triggers: this.analyzeTriggerPatterns(events || [])
    };

    if (pattern_type && patterns[pattern_type]) {
      res.json({ patterns: { [pattern_type]: patterns[pattern_type] } });
    } else {
      res.json({ patterns });
    }
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/analytics/knowledge-insights/:userId - Get knowledge graph insights
router.get('/knowledge-insights/:userId', async (req, res) => {
  try {
    const { userId } = req.params;

    const [
      { data: nodes },
      { data: relationships }
    ] = await Promise.all([
      supabase.from('knowledge_nodes').select('*').eq('user_id', userId),
      supabase.from('knowledge_relationships').select('*').eq('user_id', userId)
    ]);

    const insights = {
      graph_metrics: this.calculateGraphMetrics(nodes || [], relationships || []),
      learning_progress: this.analyzeLearningProgress(nodes || []),
      knowledge_clusters: this.identifyKnowledgeClusters(nodes || [], relationships || []),
      growth_trends: this.analyzeKnowledgeGrowth(nodes || []),
      connection_strength: this.analyzeConnectionStrength(relationships || [])
    };

    res.json({ insights });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/analytics/reports/:userId/:reportType - Generate specific reports
router.get('/reports/:userId/:reportType', async (req, res) => {
  try {
    const { userId, reportType } = req.params;
    const { startDate, endDate, format = 'json' } = req.query;

    let report;
    switch (reportType) {
      case 'productivity':
        report = await this.generateProductivityReport(userId, startDate, endDate);
        break;
      case 'learning':
        report = await this.generateLearningReport(userId, startDate, endDate);
        break;
      case 'usage':
        report = await this.generateUsageReport(userId, startDate, endDate);
        break;
      case 'ai-performance':
        report = await this.generateAIPerformanceReport(userId, startDate, endDate);
        break;
      default:
        return res.status(400).json({ error: 'Invalid report type' });
    }

    if (format === 'csv') {
      res.setHeader('Content-Type', 'text/csv');
      res.setHeader('Content-Disposition', `attachment; filename=${reportType}-report.csv`);
      res.send(this.convertToCSV(report));
    } else {
      res.json({ report });
    }
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/analytics/health-check - Database health check
router.get('/health-check', async (req, res) => {
  try {
    const { data, error } = await supabase
      .rpc('database_health_check');

    if (error) throw error;

    res.json({ health: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Helper methods (would typically be in a separate utility file)
router.getDailyCounts = function(items, days) {
  const counts = {};
  for (let i = 0; i < days; i++) {
    const date = new Date(Date.now() - i * 24 * 60 * 60 * 1000).toISOString().split('T')[0];
    counts[date] = 0;
  }

  items.forEach(item => {
    const date = new Date(item.created_at).toISOString().split('T')[0];
    if (counts.hasOwnProperty(date)) {
      counts[date]++;
    }
  });

  return counts;
};

router.calculateAvgSessionDuration = function(sessions) {
  const completedSessions = sessions.filter(s => s.session_end);
  if (completedSessions.length === 0) return 0;

  const totalDuration = completedSessions.reduce((sum, s) => {
    const duration = new Date(s.session_end) - new Date(s.session_start);
    return sum + duration;
  }, 0);

  return totalDuration / completedSessions.length / (1000 * 60); // Return in minutes
};

router.getMostActiveHours = function(sessions) {
  const hourCounts = new Array(24).fill(0);
  sessions.forEach(session => {
    const hour = new Date(session.session_start).getHours();
    hourCounts[hour]++;
  });
  return hourCounts;
};

router.getTopEventTypes = function(events) {
  const typeCounts = {};
  events.forEach(event => {
    typeCounts[event.event_type] = (typeCounts[event.event_type] || 0) + 1;
  });
  return Object.entries(typeCounts)
    .sort(([,a], [,b]) => b - a)
    .slice(0, 10)
    .map(([type, count]) => ({ type, count }));
};

router.calculateKnowledgeGrowth = function(nodes) {
  const nodesByDate = {};
  nodes.forEach(node => {
    const date = new Date(node.created_at).toISOString().split('T')[0];
    nodesByDate[date] = (nodesByDate[date] || 0) + 1;
  });
  return nodesByDate;
};

router.calculateSuggestionAccuracy = function(suggestions) {
  const responded = suggestions.filter(s => s.status !== 'pending');
  if (responded.length === 0) return 0;

  const accepted = responded.filter(s => s.status === 'accepted').length;
  return accepted / responded.length;
};

router.getTopSuggestionTypes = function(suggestions) {
  const typeCounts = {};
  suggestions.forEach(suggestion => {
    typeCounts[suggestion.suggestion_type] = (typeCounts[suggestion.suggestion_type] || 0) + 1;
  });
  return Object.entries(typeCounts)
    .sort(([,a], [,b]) => b - a)
    .slice(0, 5)
    .map(([type, count]) => ({ type, count }));
};

router.getConfidenceDistribution = function(suggestions) {
  const distribution = { low: 0, medium: 0, high: 0 };
  suggestions.forEach(suggestion => {
    const confidence = suggestion.confidence_score || 0;
    if (confidence < 0.5) distribution.low++;
    else if (confidence < 0.8) distribution.medium++;
    else distribution.high++;
  });
  return distribution;
};

router.analyzeOCRData = function(ocrLogs) {
  return {
    total_recognitions: ocrLogs.length,
    avg_confidence: ocrLogs.length ?
      ocrLogs.reduce((sum, log) => sum + (log.confidence || 0), 0) / ocrLogs.length : 0,
    text_length_stats: this.calculateTextLengthStats(ocrLogs)
  };
};

router.analyzeClickPatterns = function(clicks) {
  return {
    total_clicks: clicks.length,
    click_distribution: this.getClickDistribution(clicks),
    avg_click_frequency: this.calculateClickFrequency(clicks)
  };
};

router.analyzeMousePatterns = function(movements) {
  return {
    total_movements: movements.length,
    movement_velocity: this.calculateMovementVelocity(movements),
    movement_patterns: this.identifyMovementPatterns(movements)
  };
};

router.calculateProductivityScore = function(session) {
  const duration = session.session_end ?
    (new Date(session.session_end) - new Date(session.session_start)) / (1000 * 60) : 0;
  const ocrCount = session.ocr_logs?.length || 0;
  const clickCount = session.clicks?.length || 0;

  // Simple productivity score algorithm
  let score = 0;
  if (duration > 30) score += 30; // Active session bonus
  if (ocrCount > 10) score += 20; // OCR activity bonus
  if (clickCount > 50) score += 20; // Click activity bonus

  return Math.min(score, 100);
};

// Additional helper methods would be implemented here...

module.exports = router;

File: squire-backend/app/routers/analytics.py
"""
Analytics and Insights routes
"""
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import Optional
from uuid import UUID

from app.core.database import get_supabase, execute_rpc, execute_query, DatabaseError
from app.models.schemas import DashboardData, SessionInsights, BehaviorPatterns

router = APIRouter()


@router.get("/dashboard/{user_id}", response_model=dict)
async def get_dashboard_data(
    user_id: UUID,
    days: int = Query(30, ge=1, le=365),
    supabase=Depends(get_supabase)
):
    """Get comprehensive dashboard data"""
    try:
        from datetime import datetime, timedelta
        start_date = (datetime.now() - timedelta(days=days)).isoformat()

        # Get all data types
        sessions = await execute_query(
            table="user_sessions",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        suggestions = await execute_query(
            table="ai_suggestions",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        events = await execute_query(
            table="user_events",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        nodes = await execute_query(
            table="knowledge_nodes",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        relationships = await execute_query(
            table="knowledge_relationships",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        # Filter by date
        recent_sessions = [s for s in (sessions or []) if s.get("created_at", "") >= start_date]
        recent_suggestions = [s for s in (suggestions or []) if s.get("created_at", "") >= start_date]
        recent_events = [e for e in (events or []) if e.get("created_at", "") >= start_date]

        dashboard = {
            "overview": {
                "total_sessions": len(recent_sessions),
                "active_sessions": len([s for s in recent_sessions if not s.get("session_end")]),
                "total_suggestions": len(recent_suggestions),
                "pending_suggestions": len([s for s in recent_suggestions if s.get("status") == "pending"]),
                "total_events": len(recent_events),
                "knowledge_nodes": len(nodes or []),
                "knowledge_relationships": len(relationships or [])
            },
            "trends": {
                "daily_sessions": {},  # Would implement daily counting
                "daily_events": {},    # Would implement daily counting
                "suggestion_response_rate": (
                    len([s for s in recent_suggestions if s.get("status") != "pending"]) / len(recent_suggestions)
                    if recent_suggestions else 0
                )
            },
            "productivity": {
                "avg_session_duration": 0,  # Would calculate from session data
                "most_active_hours": [],    # Would analyze session times
                "top_event_types": [],      # Would count event types
                "knowledge_growth": {}      # Would track node creation over time
            },
            "ai_insights": {
                "suggestion_accuracy": 0,   # Would calculate from feedback
                "top_suggestion_types": [], # Would count suggestion types
                "confidence_distribution": {} # Would analyze confidence scores
            }
        }

        return {"dashboard": dashboard}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/session-insights/{session_id}", response_model=dict)
async def get_session_insights(
    session_id: UUID,
    supabase=Depends(get_supabase)
):
    """Get detailed session insights"""
    try:
        session = await execute_query(
            table="user_sessions",
            operation="select",
            filters={"id": str(session_id)},
            single=True
        )

        if not session:
            raise HTTPException(status_code=404, detail="Session not found")

        insights = await execute_rpc(
            "generate_session_insights",
            {
                "p_user_id": str(session["user_id"]),
                "p_session_id": str(session_id)
            }
        )

        return {"insights": insights}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/behavior-patterns/{user_id}", response_model=dict)
async def get_behavior_patterns(
    user_id: UUID,
    days: int = Query(30, ge=1, le=365),
    pattern_type: Optional[str] = Query(None),
    supabase=Depends(get_supabase)
):
    """Get behavioral patterns"""
    try:
        from datetime import datetime, timedelta
        start_date = (datetime.now() - timedelta(days=days)).isoformat()

        events = await execute_query(
            table="user_events",
            operation="select",
            filters={"user_id": str(user_id)},
            order_by="created_at",
            ascending=True
        )

        # Filter by date
        recent_events = [e for e in (events or []) if e.get("created_at", "") >= start_date]

        patterns = {
            "temporal": {
                "hourly": [0] * 24,
                "daily": {},
                "weekly": [0] * 7
            },
            "frequency": {},
            "sequences": {},
            "triggers": {}
        }

        # Analyze patterns (simplified implementation)
        for event in recent_events:
            # This would be more sophisticated in a real implementation
            event_type = event.get("event_type", "unknown")
            patterns["frequency"][event_type] = patterns["frequency"].get(event_type, 0) + 1

        if pattern_type and pattern_type in patterns:
            return {"patterns": {pattern_type: patterns[pattern_type]}}
        else:
            return {"patterns": patterns}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/knowledge-insights/{user_id}", response_model=dict)
async def get_knowledge_insights(
    user_id: UUID,
    supabase=Depends(get_supabase)
):
    """Get knowledge graph insights"""
    try:
        nodes = await execute_query(
            table="knowledge_nodes",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        relationships = await execute_query(
            table="knowledge_relationships",
            operation="select",
            filters={"user_id": str(user_id)}
        )

        insights = {
            "graph_metrics": {
                "total_nodes": len(nodes or []),
                "total_relationships": len(relationships or []),
                "avg_node_weight": 0,
                "avg_relationship_strength": 0
            },
            "learning_progress": {},
            "knowledge_clusters": {},
            "growth_trends": {},
            "connection_strength": {}
        }

        if nodes:
            weights = [n.get("weight", 0) for n in nodes]
            insights["graph_metrics"]["avg_node_weight"] = sum(weights) / len(weights)

        if relationships:
            strengths = [r.get("strength", 0) for r in relationships]
            insights["graph_metrics"]["avg_relationship_strength"] = sum(strengths) / len(strengths)

        return {"insights": insights}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/health-check", response_model=dict)
async def get_health_check(
    supabase=Depends(get_supabase)
):
    """Database health check"""
    try:
        health = await execute_rpc("database_health_check")
        return {"health": health}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))